### PR TITLE
Support node16 module resolution

### PR DESCRIPTION
**Pull request by remcohaszing**
## :bookmark_tabs: Summary

The `node16` module resolution requires imports to use the `.js` file extension in type definitions. This also needs to be the case for dependencies. This fixes usage of Mermaid with projects that use `"module": "node16"`.

## :straight_ruler: Design Decisions

`@rollup/plugin-typescript` is needed to make this work with the Vite setup used by Mermaid.

The module option for Mermaid internally is set to `nodenext`. This is needed to support `.json` imports. Note that setting `module` to `node16` or `nodenext` implies a matching `moduleResolution` value.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :notebook: have added documentation (if appropriate)
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by remcohaszing** [comment]
This almost fixes it, but I can’t get the Cypress tests to succeed.

**Comment by aloisklink** [comment]
We've had issues with accidentally missing the `.js` extensions before (see https://github.com/mermaid-js/mermaid/issues/4170), so having TypeScript check this for us would be really helpful.

Thanks for making this PR, I've been working on something similar (see https://github.com/mermaid-js/mermaid/pull/4217), but you managed to make a PR before I did :) My implementation didn't need `@rollup/plugin-typescript`, since Vite/TypeScript handles it for you if you use the `.ts` extension (allowed using the [`--allowImportingTsExtensions` flag in TypeScript v5.0 RC](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-rc/)).

Because you've made your PR first, you're welcome to take my commits from https://github.com/mermaid-js/mermaid/pull/4217 if helps you get the E2E and Cypress tests working.

**Comment by remcohaszing** [comment]
`allowImportingTsExtensions` is nice for application code, but it shouldn’t be used for libraries. It emits the `.ts` extension in the output, both JavaScript and type definitions, which should never be the case. TypeScript shows an error when this option is set in `tsconfig.json`, but `noEmit` is missing.

**Comment by remcohaszing** [comment]
I think the main problem with the e2e tests is that it imports TypeScript source code, which it can’t resolve anymore. The only solution I can some up with, is to test against the built version of mermaid instead of the source code.

**Comment by aloisklink** [comment]
> `allowImportingTsExtensions` is nice for application code, but it shouldn’t be used for libraries. It emits the `.ts` extension in the output, both JavaScript and type definitions, which should never be the case.

The output JavaScript is fine, because TypeScript isn't compile/emit those. Instead, Mermaid uses [Vite to transpile TypeScript files](https://vitejs.dev/guide/features.html#typescript).

However, you're right that it emits the `.ts` extension in the `.d.ts` type definitions. That seems to be a pretty major problem, since it breaks the type definitions for downstream users, unless they switch to TypeScript v5.0 too, and that would be a breaking change.

I'll close my PR!

> I think the main problem with the e2e tests is that it imports TypeScript source code, which it can’t resolve anymore. The only solution I can some up with, is to test against the built version of mermaid instead of the source code.

It might even speed up the E2E tests, since loading bundles in browsers (even on `localhost`) is normally quite a bit faster.

**Comment by remcohaszing** [comment]
> That seems to be a pretty major problem, since it breaks the type definitions for downstream users, unless they switch to TypeScript v5.0 too

Even then it would be a problem. `.ts` imports are not supposed to be published.

---

@aloisklink Am I correct that you’re not a member of the Mermaid team? It’s nice to get some feedback, but I prefer to get confirmation from the Mermaid team before I start modifying the end-to-end tests.

**Comment by aloisklink** [comment]
> @aloisklink Am I correct that you’re not a member of the Mermaid team? It’s nice to get some feedback, but I prefer to get confirmation from the Mermaid team before I start modifying the end-to-end tests.

I'm a member of the Mermaid team, but I spend more time maintaining the [mermaid-js/mermaid-cli](https://github.com/mermaid-js/mermaid-cli) project, rather than the core Mermaid project.

If you're planning on making a lot of modifications, it's probably worth getting confirmation from somebody more familiar with the E2E tests/Vite than me, e.g. @sidharthv96.

(@sidharthv96 also has a WIP PR that replaces Vite with ESBuild that may help: https://github.com/mermaid-js/mermaid/pull/4109)

We may also be able to just wait for https://github.com/vitejs/vite/issues/8993#issuecomment-1279012465 to be added to Vite. They're planning on adding `.js` -> `.ts` resolving for imports, which might fix the E2E issues you're having, but that comment was a few months ago, so we might be waiting a while.

**Comment by sidharthv96** [comment]
@aloisklink is indeed one of our very valuable core team members. 

---

Many E2E tests already use the compiled bundle via `import mermaid from './mermaid.esm.mjs';`, so adopting that shouldn't be an issue. 

---

@aloisklink, Regarding #4109, we might have to close it as we might be adding UMD support back due to https://github.com/mermaid-js/mermaid/discussions/4148.

**Comment by remcohaszing** [comment]
> @aloisklink is indeed one of our very valuable core team members.

Ah right, sorry, I misread the `Collaborator` tag as `Contributor`.

---

An example of a problematic import is https://github.com/mermaid-js/mermaid/blob/2dd906d80914b5d283e7446ef4a3110980dd849b/cypress/platform/bundle-test.js#L1.
Would it be best to wait for the UMD bundle to be restored then?

**Comment by aloisklink** [comment]
> I misread the `Collaborator` tag as `Contributor`.

Sorry! That's actually my fault, I never noticed that I had my organization visibility setting for @mermaid-js on private. It should say **(Member)** now.

(That might explain why I sometimes see @github staff members being confused that people can't see their **(Staff)** badge.)

---

> An example of a problematic import is https://github.com/mermaid-js/mermaid/blob/2dd906d80914b5d283e7446ef4a3110980dd849b/cypress/platform/bundle-test.js#L1

I think the purpose of `bundle-test.js` is to test how third-party bundles will handle importing mermaid, so it's probably best to switch that one for `import mermaid from './mermaid.core.mjs'` since that's what's listed in Mermaid's `package.json` file: https://github.com/mermaid-js/mermaid/blob/2dd906d80914b5d283e7446ef4a3110980dd849b/packages/mermaid/package.json#L11

All of the other Cypress test files can be switched to `import mermaid from './mermaid.esm.mjs'` since all the Cypress tests should support ESM.

(In case you're curious about the differences, `./mermaid.esm.mjs` should have all of Mermaid's dependencies (i.e. `lodash-es`) bundled in, so it can be used directly from a CDN. While `mermaid.core.mjs` doesn't bundle any dependencies, so that downstream users can bundle their own dependencies)

---

> Would it be best to wait for the UMD bundle to be restored then?

My gut feeling is that adding a UMD/[IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) bundle should just be a [single option in Vite](https://vitejs.dev/config/build-options.html#build-lib), so I don't think it will cause any merge conflicts with this PR. But you're welcome to wait if you want to be 100% sure there won't be any issues.

**Comment by remcohaszing** [comment]
I figured, if you really want to end to end test, you can just use `import mermaid from 'mermaid'`. That would also ensure the export map works. Unfortunately that didn’t work though, so I used `'../../packages/mermaid/dist/mermaid.esm.mjs'` everywhere.

There are also some commented imports. I decided not to touch those.

Anyway, tests passed! :smile:

---

As a downstream user I’m super happy with the bundled version of Mermaid. It allows me to run Mermaid using Puppeteer / Playwright, and then use the results in Node.js. :smile:

**Review by aloisklink — COMMENTED** [review]
Looks basically great to me!

I tested with `git switch develop && pnpm run e2e && git switch fix-remcohaszing/node16-module-resolution && pnpm run e2e` and it looks like all of the E2E test snapshots are the same too.

---

**Minor change:**, can you switch the imports in `.html` files from `import mermaid from './mermaid.esm.mjs';` instead of `import mermaid from '../../packages/mermaid/dist/mermaid.esm.mjs'`.

It's mainly just to make things consistent with all of the existing Cypress tests: https://github.com/mermaid-js/mermaid/blob/7647ae317a7b2130e32777248d25a9c4d24b8f9f/cypress/platform/class.html#L116

I believe you can do the same thing with `import exampleDiagram from '../../packages/mermaid-example-diagram/dist/mermaid-example-diagram.core.mjs';`, since it's listed in the Vite server config (see below), but no existing tests do it, so it's up to you:

https://github.com/mermaid-js/mermaid/blob/7647ae317a7b2130e32777248d25a9c4d24b8f9f/.vite/server.ts#L17-L18

---

> As a downstream user I’m super happy with the bundled version of Mermaid. It allows me to run Mermaid using Puppeteer / Playwright, and then use the results in Node.js.

I think Mermaid v10 made life a bit harder, because now the bundle is ~40 different big files, so I support adding back a single-file bundle UMD/IIFE bundle! 

By the way, we've recently added a Node.JS API to the `mermaid-cli` project for rendering Mermaid diagrams using Puppeteer if you don't want to handle all the bundling yourself, see https://github.com/mermaid-js/mermaid-cli#use-nodejs-api. We still haven't figured out a good way to get fontawesome icons bundled in offline `.svg` files, so it isn't perfect unfortunately.

File: cypress/platform/bundle-test.js
Comment (remcohaszing) [reply cypress/platform/bundle-test.js]: Unfortunately I’m not sure how to resolve that either. I went with the comment.

**Comment by remcohaszing** [comment]
> I think Mermaid v10 made life a bit harder, because now the bundle is ~40 different big files, so I support adding back a single-file bundle UMD/IIFE bundle!
> 
> By the way, we've recently added a Node.JS API to the `mermaid-cli` project for rendering Mermaid diagrams using Puppeteer if you don't want to handle all the bundling yourself, see https://github.com/mermaid-js/mermaid-cli#use-nodejs-api. We still haven't figured out a good way to get fontawesome icons bundled in offline `.svg` files, so it isn't perfect unfortunately.

I meant to say I’m happy with the ESM bundle actually. :)

Although using ESM in the browser does cause some CSP issues when using a `file://` URL. So I’ll probably switch to the UMD bundle if that becomes available again.

---

I maintain [`remark-mermaidjs`](https://github.com/remcohaszing/remark-mermaidjs) and [`gatsby-remark-mermaid`](https://github.com/remcohaszing/gatsby-remark-mermaid). I intend to update those to the latest versions of mermaid, and create a [rehype](https://github.com/rehypejs/rehype) plugin, as that is a better place for making such tranformations.

I want to support both the remark and rehype plugin in both Node.js and the browser. In order to support both, I’m going to publish a package `mermaid-isomorphic`. This will provide an an interface something like this:

```ts
interface MermaidRenderer {
  render(diagrams: string[], config: MermaidConfig): SettledPromise<string>[]
}

declare function createMermaidRenderer(): MermaidRenderer;
```

Such a factory-like function is needed to provide an efficient interface for Node.js, as it will contain logic for starting a browser when necessary, reusing it when still available, tearing it down when it’s unused, and processing multiple diagrams in one call. This logic is not provided by `mermaid-cli`. I imagine `mermaid-cli` could use this package under the hood though.

This is getting a bit off-topic for the PR shough. If this is something you are interested to support officially / cooperate on this, please let me know via Twitter DM / email or something. :)

**Review by aloisklink — APPROVED** [review]
This PR looks perfect to me now!

FYI, Sidharth's is busy for the next few weeks, but since they're the most knowledgeable on Vite/TypeScript, this PR probably won't get merged until they've had a look a it.

File: .vite/build.ts
Comment (remcohaszing) [reply .vite/build.ts:107]: Yes, according to https://vitejs.dev/config/shared-options.html#resolve-extensions, the includes allows some extensions by default, which we don’t want.

**Comment by remcohaszing** [comment]
Is there anything blocking this? The problem is it touches all imports, so the longer we wait, the more merge conflicts will arise.

**Comment by sidharthv96** [comment]
Hey @remcohaszing, If you can sync this once, I'll take a look and merge. 
I didn't find anything major to change.

**Comment by remcohaszing** [comment]
Done!

**Comment by sidharthv96** [comment]
Thank you :)

**Pull request merged**
